ext/drsyms/drsyms_elf.c

```diff
@@ -166,6 +168,47 @@ find_elf_section_by_name(Elf *elf, const char *match_name)
     return NULL;
 }
 
+/* Reads the build id into mod->build_id. */
+static void
+read_build_id(Elf *elf, elf_info_t *mod)
+{
+    Elf_Scn *scn;
+    for (scn = elf_getscn(elf, 0); scn != NULL; scn = elf_nextscn(elf, scn)) {
+        Elf_Shdr *section_header = elf_getshdr(scn);
+        if (section_header == NULL || section_header->sh_type != SHT_NOTE)
+            continue;
+        Elf_Data *data = elf_getdata(scn, NULL);
+        Elf_Note *note = (Elf_Note *)data->d_buf;
+        if (note->n_type == NT_GNU_BUILD_ID) {
+            /* Following the note are the name and value. */
+            byte *src = ((byte *)(note + 1)) + note->n_namesz;
+            size_t size = note->n_descsz;
+            if ((byte *)data->d_buf + data->d_size < src + note->n_descsz) {
+                NOTIFY_ELF("note data is shorter than specified length");
+                size = (byte *)data->d_buf + data->d_size - src;
+            }
+            char *dst = mod->build_id;
+            for (int i = 0; i < size; i++) {
```

Comment:
A couple of things in this loop took me a while to understand.
1. The +3 in the size check wasn't apparent until I read the dr_snprintf call below.
2. NULL_TERMINATE_BUFFER or similar is not required since each time dr_snprintf writes out 2 chars for the actual value and the null terminator. Then we overwrite the null terminator the next time around.

Perhaps a couple of comments would help here?

Reply:
Added an explanation of the "3".  Removed the NULL_TERMINATE and added a comment; as you said with the up-front overflow check it is not needed.